### PR TITLE
Draft practices for versioning packages

### PR DESCRIPTION
**Pull request by jakebeal**
I have distilled some discussions from iGEM on how to version packages given the length time between ordering designs and finding out if they can be produced.

### CODE REVIEW

**Comment by noahsprent** [comment]
Thanks Jake! It's an excellent summary of our discussion, and I think makes sense for our workflow in the distribution for now. I still feel like we're missing a few things for the more general cases, or at least I'm missing understanding a few things . For example, how does one distinguish between software and wetware packages? If one sees that a release is a release candidate, then they can infer from that that the maintainers are currently trying to build the package in wetware. But how is one to know when seeing a non-rc whether this means the package has been physically built or just that the package isn't intended to be built? And what if certain parts in the package have been built but others have not?

**Comment by noahsprent** [comment]
Again, just throwing out thoughts here, but I'm wondering if maybe representing the physical status of packages needs to be done in a way more akin to testing rather than releasing. In the same way that one can release a version of a software package that is failing tests (although it isn't recommended), one could release a version of an SBOL package where not all of the parts have been synthesised. The repo would have a badge saying "53% synthesised" or something like that. Tracking of whether or not a part had been synthesised or not could be done using the existing SBOL structure of implementations. All the 'test' does is see which parts have implementations. Some packages could then choose to use the 'release-candidate->release' method to ensure that they only ever release with 100% build coverage, but for those packages which don't intend to be built, or ones where they contain parts that shouldn't be built, one can quickly see which parts have been built and which haven't.

File: sep_054.md [hunk 1]
```diff
@@ -370,6 +371,39 @@ Package build for distribution:
...
+Note that these policies apply only to designs in a package, not to its dependencies.
+As a matter of practice, however, a production release generally SHOULD NOT have any dependency on any pre-release packages.
+
+##### Example:
+
+Consider an `ecoli-circuits` package at version `1.0`, to which a number of new designs have been added with the intention of producing a `1.1` release.
+When design work is complete, including a check that the designs are expected to be able to be synthesized, the package is tagged for pre-release version `1.1-rc1` and the new designs are sent for synthesis.
+
+Several weeks later, most of the designs have been synthesized successfully, but a few have failed.
+The failing designs are updated to try to resolve their synthesis issues. The package is then tagged as pre-release version `1.1-rc2` and the updated designs are sent for synthesis.
```
Comment (vinoo-igem) [inline sep_054.md:396]: Not being too familiar with semver, would this be an area why we should consider incrementing a patch number? 1.1-rc1 to 1.1.1-rc1 or is the rc# taking that place?  

I think this highlights an area of confusion for me, as reevaluating a design (not just reordering the previous one) for production, seems like a bugfix. There's also an assumption here that the updated design will perform the same as the previous one, so it can replace it.

On the other hand there also be an assumption that all synthesized products work as intended, were created as intended, or should be distributed/available. So instead, after release, how do we account for bugfixing a package.

In short, in what cases and at what stages are packages fixed and where does versioning reflect that?

Comment (noahsprent) [reply sep_054.md:396]: I think you raise some good points. I would say that the patch number would increment if we're changing for knowledge/software reasons, rc number for build reasons. But then what if you add a logical bugfix to a part e.g. fixing a mutation and add that as a patch. Can that release then only be a release candidate until every part in it has been synthesised? Definitely is easier if we just ignore the physical implementation of parts and just focus on the software, but I do feel like that is really ignoring what makes biology fundamentally different from software.

**Comment by jakebeal** [comment]
@noahsprent I think that the tracking and build badges are a great idea, but I think we should go through the exercise a couple of times and experiment with it before we add it to the SEP, since I'm not quite sure what's going to make sense in terms of keeping that information in the package vs. in build artifacts vs. in external databases. I've added a note trying to capture the idea as future work in the discussion section.

I likewise think that we should leave the "software vs. wetware" undefined for the moment, since right now everything we're doing is wetware-focused.

File: sep_054.md (hunk 1 above)
Comment (jakebeal) [reply sep_054.md:396]: The key distinction here is that an "rc" version is a pre-release, which tells both people and software that its contents are unstable and cannot be counted on to remain the same until the release is finalized. Thus, if we run into problems (logical or physical) with a pre-release, we can fix them without affecting the major/minor version number.  

What this means is that if we find out that part `SuperHairpin` in `1.1-rc1` cannot be synthesized, then we can just drop it and release `1.1` with `SuperHairpin` having been deleted, since `1.1` will still contain all of the materials from `1.0` and thus be backward-compatible. But if we release `1.1` before `SuperHairpin` fails, then the new release that drops it would have to be `2.0` because it is not backward-compatible, having removed materials that were made available in `1.1`.

I've added a note in the discussion section to try to express this reasoning.

**Comment by noahsprent** [comment]
@jakebeal Good point! No point trying to optimise everything too early. It would be nice in future workflows to be able to distinguish between `sip import`ing a physical part to represent someone sending you a strain. e.g. iGEM teams from the distribution, vs a design that someone has created e.g. the distro importing designs from other packages to then synthesise. But again, I'm getting overexcited and that's for the future!

**Comment by jakebeal** [comment]
Merging per discussion with @noahsprent and @vinoo-igem
